core/src/consensus.rs

```diff
@@ -515,6 +515,60 @@ impl Tower {
             .map(|last_voted_slot| {
                 let root = self.root();
                 let empty_ancestors = HashSet::default();
+                let empty_ancestors_due_to_minor_unsynced_ledger = || {
```

Comment:
nice, I like this new structure! I think if possible we should add a simple diagram/explanation to the comment to help clarify these points/add proof to the correctness. 

Maybe we can import from here
1) Safety: https://github.com/solana-labs/solana/pull/10718#discussion_r474468319

2) Corner case consideration: https://github.com/solana-labs/solana/pull/10718/files#r478740859

Reply:
Yeah, I'll do this later: https://github.com/solana-labs/solana/pull/12671#issuecomment-719473558